Add TimeFormatter class

teggotic opened the pull request:
Fixes https://github.com/jemc/mare/issues/2

teggotic commented:
But I'm stuck coz I can't compile this package now. 
I'm getting this issue
![image](https://user-images.githubusercontent.com/16068886/86976741-cb113300-c183-11ea-9e21-2d8bbec36ef0.png)
when compiling 
```
:import "time"

:actor Main
  :new(env Env)
    env.out.print("Hello")
```

teggotic commented:
I found issue with syntax. We need to improve error messages :upside_down_face:

teggotic commented:
I'm stuck now with new issue. Also there is no way to get any info about it.
![image](https://user-images.githubusercontent.com/16068886/87087160-484baf00-c23b-11ea-9129-b193722c7a93.png)
As far as I understand, mare cant infer type for a number literal and so it cant compile generate ir for it.

jemc commented:
@teggotic - Thanks for working on this, and thanks for your other issue ticket and pull request contributions.

If you're interested, I'd love to have a live pairing session with you so I could answer more of your questions more quickly - feel free to DM me in the Pony Zulip if you'd like to set up a session.

You've correctly diagnosed this particular numeric literal issue, so you just need the right syntax for it - the answer is to use the `U64[]` method.

For example:
```mare
number = U64[0]
```

I will do an update soon to make sure we have a nice, helpful error message for that error and all the other ones you run into.

Again, I appreciate your contributions these last few days and I will be responding to each of them soon. Thanks!

jemc commented on packages/time/formatter.mare:
```diff
@@ -0,0 +1,504 @@
+  :new (@pattern)
+
+  :fun format (time Time'box)
+    i USize = 0
```
```suggestion
    i USize = U64[0]
```

jemc commented:
I've added a nice error message for the numeric literal issue in https://github.com/jemc/mare/commit/bfefb32ecd97bb754de101b540b1c809cb4749c6

Thanks for bringing that case to my attention.

teggotic commented:
> I've added a nice error message for the numeric literal issue in [bfefb32](https://github.com/jemc/mare/commit/bfefb32ecd97bb754de101b540b1c809cb4749c6)
> 
> Thanks for bringing that case to my attention.

This is excellent. I fixed my issue in 10 minutes

teggotic commented:
I've stuck with new issue. 
![image](https://user-images.githubusercontent.com/16068886/87175036-beeeb800-c2e0-11ea-879b-302518d92437.png)

jemc commented on packages/time/formatter.mare:
```diff
@@ -0,0 +1,517 @@
+:class TimeFormatter
+  :const month_names Array(String): [
```
The compiler doesn't support complex types for constants yet, so this needs to be a `:fun non` for now, until I add that feature. This is the cause of the compiler bug you observed.

```suggestion
  :fun non month_names Array(String): [ // TODO: make const when array constants are supported
```

The performance will be better when it can be a `:const`, but the code should work fine as a `:fun non` for now.

jemc commented on packages/time/formatter.mare:
```diff
@@ -0,0 +1,517 @@
+    "December"
+  ]
+
+  :const day_names Array(String): [
```
```suggestion
  :fun non day_names Array(String): [ // TODO: make const when array constants are supported
```

jemc commented:
Thanks for working on this. I will take the time to review this in depth tomorrow.

jemc commented on src/prelude/string.mare:
```diff
@@ -1,3 +1,4 @@
+:import "collections" (Count)
```
I think I want to refactor this to not use `Count` - just because I don't want the `prelude` to have a dependency on any other package, including the `collections` package.

But I can do that refactor myself after merging.

Thanks for all of this!

jemc replied:
It turns out that having this dependency is causing other parts of the test suite in the CI run to break.

So I'm going to commit my fix to this branch, simultaneously refactoring `String.substring` to use the more performant unsafe CPointer operations.
